### DISCUSSION TITLE
Use the pinocchio.casadi module inside an existing docker container

### DISCUSSION THREAD

**Discussion by FrancescoD98** [stack-of-tasks/pinocchio#2395] [Q&A]
Hi, first of all thank you for your amazing work.

I'm currently working on a project on ROS noetic inside an existing docker container using python.
I would like to use the new casadi support of pinocchio 3, however I've tried to install the latest version using (1) the [installation guide](https://stack-of-tasks.github.io/pinocchio/download.html), (2) building it from source and (3) using conda.

 - With the first method python doesn't find the module `pinocchio.casadi`. The installation procedure was to copy-paste exactly the command on the download page.
 - When building from source, the error is
```
>>> import pinocchio.casadi
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/openrobots/lib/python3.8/site-packages/pinocchio/casadi/__init__.py", line 11, in <module>
    from ..pinocchio_pywrap_casadi import *
ImportError: libcasadi.so.3.7: cannot open shared object file: No such file or directory
```
`pinocchio` was built with the options `-DCMAKE_INSTALL_PREFIX=/opt/openrobots` and `-DBUILD_WITH_CASADI_SUPPORT=ON ` 
 - finally the installation with `conda` works outside the container

Is there anything I'm missing? Could you please provide instructions on how to install everything as it should be?

Thanks in advance for your kind reply.

**Comment by caineblood** [comment]
The above comment asking you to download a file is malware to steal your account; do not under any circumstances download or run it. The post needs to be removed. If you have attempted to run it please have your system cleaned and your account secured immediately.

**Comment by jcarpent** [comment]
As far as I know (I'm not a ROS expert), the Python CasADi module is only available on conda packager manager.
It should still be included into ROS, apt, pip, homebrew package managers.

**Reply by FrancescoD98** [reply]
ROS noetic forces me to use docker, since it only works in ubuntu 20, while my os is 22.
I tried to install pinocchio with conda from inside and outside the container, but I cannot import pinocchio from inside the container in any case.

**Reply by nim65s** [reply]
Packages from conda and from apt are not compatible. 

Apt packages from ubuntu, robotpkg and ROS are compatible, but some overlap are incompatible, so you must use either ros-noetic-eigenpy, ros-noetic-hpp-fcl & ros-noetic-pinocchio or robotpkg-py38-eigenpy, robotpkg-py38-hpp-fcl & robotpkg-py38-pinocchio.

Apt packages from ROS do not provide casadi (ref. https://index.ros.org/p/casadi/)

Apt packages from robotpkg should include casadi, but according to your error this seems broken right now. I will investigate this and push a fix, but this probably won't be ready until next week.

**Reply by FrancescoD98** [reply]
@nim65s thank you a lot. I'm looking forward to the fix.

To recreate the same conditions, I'm using [this guide](https://github.com/DIAG-Robotics-Lab/labrob_tiago_docker) to build new containers using the base docker image of TIAGo reported [here](http://wiki.ros.org/Robots/TIAGo/Tutorials/Installation/Installing_Tiago_tutorial_docker).

Don't know if this can be helpful, however when building from source inside the container and running `python -v -c "import pinocchio; import pinocchio.casadi"`, the error seems to be the following one
```
# extension module 'pinocchio.pinocchio_pywrap_default' loaded from '/usr/local/lib/python3.8/site-packages/pinocchio/pinocchio_pywrap_default.cpython-38-x86_64-linux-gnu.so'
# extension module 'pinocchio.pinocchio_pywrap_default' executed from '/usr/local/lib/python3.8/site-packages/pinocchio/pinocchio_pywrap_default.cpython-38-x86_64-linux-gnu.so'
import 'pinocchio.pinocchio_pywrap_default' # <_frozen_importlib_external.ExtensionFileLoader object at 0x729e81f33430>
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/utils.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/__pycache__': PermissionError(13, 'Permission denied')
import 'pinocchio.utils' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a5a71f0>
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/explog.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/__pycache__': PermissionError(13, 'Permission denied')
import 'pinocchio.explog' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a5a74f0>
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/robot_wrapper.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/__pycache__': PermissionError(13, 'Permission denied')
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/shortcuts.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/__pycache__': PermissionError(13, 'Permission denied')
# /usr/lib/python3.8/__pycache__/typing.cpython-38.pyc matches /usr/lib/python3.8/typing.py
# code object from '/usr/lib/python3.8/__pycache__/typing.cpython-38.pyc'
import 'typing' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a5b9820>
import 'pinocchio.shortcuts' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a5b9550>
import 'pinocchio.robot_wrapper' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a5a78b0>
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/deprecated.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/__pycache__': PermissionError(13, 'Permission denied')
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/deprecation.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/__pycache__': PermissionError(13, 'Permission denied')
import 'pinocchio.deprecation' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a541e50>
import 'pinocchio.deprecated' # <_frozen_importlib_external.SourceFileLoader object at 0x729e7a5b9970>
import 'pinocchio' # <_frozen_importlib_external.SourceFileLoader object at 0x729e81f86eb0>
# code object from /usr/local/lib/python3.8/site-packages/pinocchio/casadi/__init__.py
# could not create '/usr/local/lib/python3.8/site-packages/pinocchio/casadi/__pycache__': PermissionError(13, 'Permission denied')
# /home/alienware/.local/lib/python3.8/site-packages/casadi/__pycache__/__init__.cpython-38.pyc matches /home/alienware/.local/lib/python3.8/site-packages/casadi/__init__.py
# code object from '/home/alienware/.local/lib/python3.8/site-packages/casadi/__pycache__/__init__.cpython-38.pyc'
# /home/alienware/.local/lib/python3.8/site-packages/casadi/__pycache__/casadi.cpython-38.pyc matches /home/alienware/.local/lib/python3.8/site-packages/casadi/casadi.py
# code object from '/home/alienware/.local/lib/python3.8/site-packages/casadi/__pycache__/casadi.cpython-38.pyc'
# destroy casadi.casadi
# destroy casadi
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/usr/local/lib/python3.8/site-packages/pinocchio/casadi/__init__.py", line 5, in <module>
    from ..pinocchio_pywrap_casadi import *
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 657, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 556, in module_from_spec
  File "<frozen importlib._bootstrap_external>", line 1166, in create_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
SystemError: initialization of pinocchio_pywrap_casadi raised unreported exception

```

**Reply by FrancescoD98** [reply]
Hi @nim65s, are there any news about this topic?

**Reply by nim65s** [reply]
I made some progress, but not enough yet. And I don't think I'll be able to finish it this week. Probably the next one :)
